mjpc/tasks/humanoid/tracking/task.cc

```diff
@@ -0,0 +1,157 @@
+      int body_mocapid = model->body_mocapid[
+        mj_name2id(model, mjOBJ_BODY, mocap_body_name.c_str())];
+      assert(0 <= body_mocapid);
+
+      double mocap_body_linvel[3];
+      mju_copy3(mocap_body_linvel, data->mocap_quat + 3 * body_mocapid);
+      // double current_mocap_body_pos[3];
+      // mju_copy3(current_mocap_body_pos, model->key_mpos + model->nmocap * 3 * (step_index + 0) + 3 * body_mocapid);
+      // double next_mocap_body_pos[3];
+      // mju_copy3(next_mocap_body_pos, model->key_mpos + model->nmocap * 3 * (step_index + 1) + 3 * body_mocapid);
+      // mju_sub3(mocap_body_linvel, next_mocap_body_pos, current_mocap_body_pos);
+      // mju_scl3(mocap_body_linvel, mocap_body_linvel, fps);
+
+      double* sensor_linvel = mjpc::SensorByName(model, data, linvel_sensor_name.c_str());
+
+      mju_sub3(&residual[counter],
+               mocap_body_linvel,
+               sensor_linvel);
+      counter += 3;
+    }
+
+    // sensor dim sanity check
+    // TODO: use this pattern everywhere and make this a utility function
+    int user_sensor_dim = 0;
+    for (int i=0; i < model->nsensor; i++) {
+      if (model->sensor_type[i] == mjSENS_USER) {
+        user_sensor_dim += model->sensor_dim[i];
+      }
+    }
+    if (user_sensor_dim != counter) {
+      std::printf("user_sensor_dim=%d, counter=%d", user_sensor_dim, counter);
+      mju_error_i("mismatch between total user-sensor dimension"
+                  "and actual length of residual %d", user_sensor_dim);
+    }
+
+  }
+
+  // -------- Transition for humanoid task ---------
+  //   TODO(hartikainen)
+  // -----------------------------------------------
+  int Humanoid::Tracking::Transition(int state, const mjModel* model, mjData* data) {
+    // TODO(hartikainen): Add distance-based target transition logic.
+    // TODO(hartikainen): is `data->time` the right thing to index here?
+    float fps = 30.0;
+
+    // Positions:
+    // Linearly interpolate between two consecutive key frames in order to
+    // provide smoother signal for tracking.
+    int last_key_index = (model->nkey) - 1;
+    int key_index_0 = std::clamp((data->time * fps), 0.0, (double)last_key_index);
+    int key_index_1 = std::min(key_index_0 + 1, last_key_index);
+
+    double weight_1 = std::clamp(data->time * fps, 0.0, (double)last_key_index) - key_index_0;
+    double weight_0 = 1.0 - weight_1;
+
+    double mocap_pos_0[3 * model->nmocap];
+    double mocap_pos_1[3 * model->nmocap];
+
+    mju_scl(mocap_pos_0,
+            model->key_mpos + model->nmocap * 3 * key_index_0,
+            weight_0,
+            model->nmocap * 3);
+
+    mju_scl(mocap_pos_1,
+            model->key_mpos + model->nmocap * 3 * key_index_1,
+            weight_1,
+            model->nmocap * 3);
+
+    mju_copy(data->mocap_pos, mocap_pos_0, model->nmocap * 3);
+    mju_addTo(data->mocap_pos, mocap_pos_1, model->nmocap * 3);
+
+    // Velocities:
+    mju_copy(data->mocap_quat,
```

Comment:
This is misusing `mocap_quat` and the velocities should probably be computed in residual instead but that fails for whatever reason. See my comment above.